[Quirk] Allow pinterest.com to request push notifications without user gesture

vitorroriz opened the pull request:
#### 190dac6ff2e2090ebc0738141036741ff3caea2c
<pre>
[Quirk] Allow pinterest.com to request push notifications without user gesture
<a href="https://bugs.webkit.org/show_bug.cgi?id=303793">https://bugs.webkit.org/show_bug.cgi?id=303793</a>
<a href="https://rdar.apple.com/104979314">rdar://104979314</a>

Reviewed by Brent Fulgham.

Pinterest calls pushManager.subscribe() after page navigation, which lacks a user
gesture in the document. WebKit blocks this type of request while other engines
don&apos;t (Blink only track it for telemetry).

Add a quirk for pinterest.com to allow push notification subscription without user gesture,
matching other engines behavior for this site.

This quirk should be removed once Pinterest updates their implementation to request
notifications from a user-triggered action (e.g., clicking a button).

* Source/WebCore/Modules/push-api/PushManager.cpp:
(WebCore::PushManager::subscribe):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldAllowNotificationPermissionWithoutUserGesture const):
(WebCore::handlePinterestQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: <a href="https://commits.webkit.org/304205@main">https://commits.webkit.org/304205@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c2ab22d12a98c827d4d4ce4f18f638cc8de73f24

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134899 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7318 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46140 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142406 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4b0dc5c0-d58b-4bf6-8c3a-db5a9cb4b7cd) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136768 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7938 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7165 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103071 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/57303b9a-f044-4115-875c-de5f384e3252) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137845 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5607 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120906 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83918 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5434 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3046 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3000 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39062 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145105 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6994 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39637 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7049 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5865 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111805 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28363 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5269 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117189 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60919 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7042 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35362 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6820 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70617 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7057 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6930 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -3166,6 +3164,15 @@ static void handlePandoraQuirks(QuirksData& quirksData, const URL& /* quirksURL
+    if (quirksDomainString != "pinterest.com"_s) [[unlikely]]
+        return;
+
+    // pinterest.com rdar://104979314
```
Can you also add a comment that this quirk should be removed when rdar://165745719 is resolved?

brentfulgham reviewed (approved):
This looks great! Please add a "FIXME" comment (regarding rdar://165745719) to remove this quirk once that radar is resolved.

karlcow reviewed (approved):
approved with the comment of brent.

vitorroriz committed:
[Quirk] Allow pinterest.com to request push notifications without user gesture
https://bugs.webkit.org/show_bug.cgi?id=303793
rdar://104979314

Reviewed by Brent Fulgham.

Pinterest calls pushManager.subscribe() after page navigation, which lacks a user
gesture in the document. WebKit blocks this type of request while other engines
don't (Blink only track it for telemetry).

Add a quirk for pinterest.com to allow push notification subscription without user gesture,
matching other engines behavior for this site.

This quirk should be removed once Pinterest updates their implementation to request
notifications from a user-triggered action (e.g., clicking a button).

* Source/WebCore/Modules/push-api/PushManager.cpp:
(WebCore::PushManager::subscribe):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldAllowNotificationPermissionWithoutUserGesture const):
(WebCore::handlePinterestQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: https://commits.webkit.org/304205@main
